File: q151-175/q160.py
# 2021.04.06
# Problem Statement:
# https://leetcode.com/problems/intersection-of-two-linked-lists/
# I clearly remember I came across this problem in my ECE244 final exam, which was in Dec, 2018.
# I felt so bad that I still didn't know how to solve it...
# https://leetcode.com/problems/intersection-of-two-linked-lists/discuss/49785/Java-solution-without-knowing-the-difference-in-len!

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        # check empty linkedlist
        if not headA or not headB: return None
        
        # initialize pointers, each will travel at the same pace, 
        # and swtich to another list when the current list is done
        ptr_a, ptr_b = headA, headB
        
        # make sure it will not be set to headA or headB multiple times (only once) 
        loop_once = False
        
        while True:
            if ptr_a == ptr_b:
                return ptr_a
            else:
                # if not reaches the end                
                if ptr_a.next:
                    ptr_a = ptr_a.next
                # if reaches the end and hasn't looped once                              
                elif not ptr_a.next and not loop_once:
                    ptr_a = headB
                    # set the flag, cannot loop again
                    loop_once = True
                else:
                    # no intersection is found                    
                    return None
                
                # if not reaches the end
                if ptr_b.next:
                    ptr_b = ptr_b.next
                # if reaches the end (loop_once for ptr_a would take care of everything)
                else:
                    ptr_b = headA

File: Amazon/q160.java
// 2022.06.29
// Problem Statement:
// https://leetcode.com/problems/intersection-of-two-linked-lists/

// idea: get the length difference of the 2 linkedlists and move the longer one by difference,
// then move 2 pointers together, until they reach an intersection
// another idea without knowing the length difference:
// https://leetcode.com/problems/intersection-of-two-linked-lists/discuss/49785/Java-solution-without-knowing-the-difference-in-len!
// in first iteration, reset pointer to another's head if reaches the tail,
// so the next time they both reach to the tail at the same time, as both of them travels m+n nodes,
// but do an early stop when they intersect, as the intersection part is identical 
// and won't introduce difference in travel time
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int length1=0, length2=0;
        ListNode ptr1=headA, ptr2=headB;
        while (ptr1!=null) {
            length1++;
            ptr1 = ptr1.next;
        }
        while (ptr2!=null) {
            length2++;
            ptr2 = ptr2.next;
        }
        // reset 2 pointers
        ptr1=headA;
        ptr2=headB;
        // move the longer one by difference
        if (length1>length2) {
            for (int i=0; i<length1-length2; i++) {
                ptr1 = ptr1.next;
            }
        } else {
            for (int i=0; i<length2-length1; i++) {
                ptr2 = ptr2.next;
            }
        }
        
        while (ptr1!=ptr2) {
            ptr1 = ptr1.next;
            ptr2 = ptr2.next;
        }
        return ptr1;
    }
}